packages/compass-import-export/src/utils/collection-stream.js

```diff
@@ -69,57 +71,85 @@ class WritableCollectionStream extends Writable {
     this._executeBatch(callback);
   }
 
-  _executeBatch(callback) {
+  async _executeBatch(callback) {
     const documents = this.batch;
 
     this.batch = [];
 
-    this._collection().bulkWrite(
-      documents.map((doc) => ({ insertOne: doc })),
-      {
-        ordered: this.stopOnErrors,
-        retryWrites: false,
-        checkKeys: false,
-      },
-      (err, res) => {
+    const bulkWrite = promisify(
+      this._collection().bulkWrite.bind(this._collection())
+    );
+    let result;
+    let error;
+
+    try {
+      result = await bulkWrite(
+        documents.map((doc) => ({ insertOne: doc })),
+        {
+          ordered: this.stopOnErrors,
+          retryWrites: false,
+          checkKeys: false,
+        }
+      );
+    } catch (bulkWriteError) {
+      // Currently, the server does not support batched inserts for FLE2:
+      // https://jira.mongodb.org/browse/SERVER-66315
+      // We check for this specific error and re-try inserting documents one by one.
+      if (bulkWriteError.code === 6371202) {
+        const insertOne = promisify(
+          this._collection().insertOne.bind(this._collection())
+        );
+
+        try {
+          let nInserted = 0;
+          await Promise.allSettled(
+            documents.map(async (doc) => {
+              try {
+                await insertOne(doc, {});
+                nInserted += 1;
+              } catch (error) {
+                this._errors.push(error);
+              }
+            })
+          );
```

Comment:
This would currently run in parallel, which is nice for performance, but I’m not sure how bad it is to potentially insert in an order which doesn’t match the import source file’s order, or to not stop the insert if `stopOnErrors` is set and an error occurs on one of the inserts.

Maybe this could just be a plain `for` loop? I think the performance impact here would be acceptable either way, it’s fine for this to be slow, especially if the server will add support for batched writes sometime in the future anyway.

Reply:
Ths for pointing at this! Post this here for reference: https://www.mongodb.com/docs/manual/core/bulk-write-operations/#ordered-vs-unordered-operations